### DISCUSSION TITLE
Executing preprocessor in RAMSECTIONs?

### DISCUSSION THREAD

**Discussion by Pokeman2003** [vhelin/wla-dx#634] [Q&A] [answered]
So, I've made a file which kind of looks like this. ~~(I'm aware the behavior I'm attempting to achieve here is already covered, but I just wanted to use this for demonstration anyways ᵃⁿᵈ ᵗʰᶦˢ ᶦˢ ᵗᵒᵗᵃˡˡʸ ⁿᵒᵗ ᵐʸ ᶠᵃᶦˡᵉᵈ ᵃᵗᵗᵉᵐᵖᵗˢ ᵗᵒ ᵃᶜʰᶦᵉᵛᵉ ˢᵒᵐᵉᵗʰᶦⁿᵍ ᵗʰᵃᵗ ᶜᵒᵘˡᵈ'ᵛᵉ ᵇᵉᵉⁿ ᶠᵒᵘⁿᵈ ᶦⁿ ᵗʰᵉ ᵈᵒᶜᵘᵐᵉⁿᵗᵃᵗᶦᵒⁿ ʰᵃᵈ ᴵ ˡᵒᵒᵏᵉᵈ ᶠᵒʳ ˡᵒⁿᵍᵉʳ ᵗʰᵃⁿ ¹² ˢᵉᶜᵒⁿᵈˢ)~~
```asm
.include "SYS.map"
.DEFINE STACK_SIZE 8
.DEFINE STACK_START $00FF-(STACK_SIZE*4)	; Adjusted to 4 for testing purposes, it's meant to be 2.

.MACRO STACK_GEN
	.REPEAT STACK_SIZE
		.	dw
	.ENDR
.ENDM

.PRINT "Reserved ", HEX (STACK_SIZE*2), " bytes for stack, stored @ ", HEX STACK_GEN, ". (", HEX STACK_SIZE, " values)\r\n"
.RAMSECTION "Stack" SLOT 1 FORCE ORGA STACK_START ; ORG throws a linker error about not having any space in bank 0, by the ways.
	STACK_GEN			; This does not work.
	.REPEAT STACK_SIZE	; This will also throw an assembling error.
		.	dw
	.ENDR
.ENDS
```
I believe the issue is that both of these treat the preprocessor instructions STACK_GEN and .REPEAT as names, which is a bit annoying. I'm hoping there's some way to prevent this behavior, such as a behavioral escape character which doesn't appear to have been mentioned in documentation. 
If one doesn't exist, then this should probably be implemented when someone has a chance, since I can already imagine a couple of fringe cases where using macros in a RAMSECTION could be handy.

For expedient testing, my absolute behemoth of a [SYS.map file](https://github.com/user-attachments/files/16540953/SYS.txt) is provided.
Edit: Never mind, the behavior I wanted to achieve(to automatically fill a RAMSECTION up with 16 reserved bytes using a define) was not achieved the way I was hoping. That makes a macro or a repeat even more desirable to me.

**Comment by vhelin** [comment, marked as answer]
Just a quick note: .MACRO and .REPEAT are not processed by the preprocessor, they are traversed when parsing the code. Also: .RAMSECTIONs accept only variable labels and variable sizes (see .ENUM) so you cannot use .MACRO or .REPEAT inside them.

I'm not 100% I understand what you want to achieve here, but couldn't you do something like

.RAMSECTION "Stack" SLOT 1 FORCE ORGA STACK_START
  .   dsw STACK_SIZE
.ENDS

?

**Reply by Pokeman2003** [reply]
I'm trying to make a redefinable stack. This way I can easily adjust the stack size if it turns out that I underestimated how big the stack should be.
DSW works flawlessly for this, and also works perfectly with my other use-case of creating a spot in RAM to store my characters. This is what I like about you, despite my inherent stupidity you're always willing to help me out! Now, to figure out how to do decimal math in defines... (IE: (2/4)*(10\*10) = 100, not 50 like I'd expect. I suspect I'm just going to have to forgo that.)
Edit: Nevermind it worked flawlessly, I just had a w where there should've been a b. No duh that's going to take 120 bytes instead of 60.

**Comment by vhelin** [comment]
Also not sure (cannot test code right now), but a workaround where you create .RAMSECTIONs (pieces of the big one you want to create) inside a .MACRO/.REPEAT and then use APPENDTO to join the pieces might also work?

**Reply by Pokeman2003** [reply]
I'll experiment with this potential workaround later, since this may resolve a completely different "problem" that may come up thanks to the DSW solution.
